File: common/src/main/java/dev/obscuria/fragmentum/common/signal/_Signal2.java
package dev.obscuria.fragmentum.common.signal;

import dev.obscuria.fragmentum.v2.api.common.signal.Signal2;

public final class _Signal2<P1, P2> extends _Signal<Signal2.Listener<P1, P2>> implements Signal2<P1, P2> {

    @Override public void emit(P1 p1, P2 p2) {
        emitInternal(it -> it.consume(p1, p2));
    }
}

File: api/src/main/kotlin/dev/obscuria/fragmentum/tools/signal/Signal2.kt
@file:Suppress("unused", "UNUSED_PARAMETER", "RedundantNullableReturnType")

package dev.obscuria.fragmentum.tools.signal

class Signal2<P1, P2> : Signal<Signal2.Listener<P1, P2>>() {

    fun emit(p1: P1, p2: P2): Unit = error("API stub")

    fun interface Listener<P1, P2> {

        fun consume(p1: P1, p2: P2)
    }
}